enable backward pass computation and communication overlap by prefetching all gather

zhaojuanmao opened the pull request:
Summary:
address comments in https://github.com/pytorch/pytorch/pull/69282:
Have fixed a few corner cases for prefetching full parameters in post backward hook.

After benchmarking, prefetching full parameters in the pre-backward hook has the best performance and stable but at cost of increased memory; prefetching full parameters in the post-backward hook did not see expected performance, also failed in a few corner cases (fixed) although there is no memory increase. The main issue is that post backward hook fire order is not consistent with opposite of forward computation order, so incorrectly prefetched all gather could delay the really needed all gather in the single NCCL stream and cause some layer's computation delay.

So putting  these two algorithms as two configurable experimental algorithms for now

prefetch full parameters at pre-backward hook:

It is observed from past traces that all gather ops are not triggered until current layer's backward pass starts to compute, also for some models previous layers' reduce scatter is scheduled before next layer's all gather ops, since all gather and reduce scatter are in the same nccl stream, this case could result in backward pass has no communication and computation overlap.

To explicitly make next layers' all gather scheduled while previous layers' backward computation is running, we can prefetch next layers' all gather full params. This can help 1) both all gather and reduce scatter are overlapped with computation deterministically 2) only prefetch one layer's all gather full parameters, to avoid increasing too much memories.

The implementation borrowed the idea from facebookresearch/fairscale#865, where forward graph order is recorded in the forward pass.

In the backward pass, this PR prefetches all gather full parameter in current layer's pre-backward hook, instead of prefetching in current layer's post backward hook in facebookresearch/fairscale#865. Also make sure all gather streams are synced properly.

Experiments showed 10% memory increase and 20% latency speed up for 1GB roberta model in a slow network environment.

Test Plan: unit tests

Differential Revision: D33252795



cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

rohan-varma commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -701,9 +766,26 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
+            self._fsdp_graph_order[self._my_fsdp_idx_in_graph - 1]._rebuild_full_params()  # type: ignore[operator]
+            # Next layer's computation will start right after this all_gather,
+            # Wait for all_gather to finish before computation.
+            torch.cuda.current_stream().wait_stream(self._streams["all_gather"])
```
this is because the next layer's pre-backward hook might already be running? If we had guaranteed order such that post backward hook for this layer comes before pre-backward hook for next layer, we wouldn't need this wait?

zhaojuanmao replied:
next layer's computation will start immediately, so wait for here. it is like moving up _rebuild_full_params in pre-backward hook here, this move up is to make all gather scheduled before reduce_scatter, so that reduce_scatter is not blocked and can be overlapped with computation

rohan-varma reviewed (approved):
Awesome and thank you for the thorough verification and bugfixes! Have a few questions but overall LGTM

rohan-varma commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -701,9 +766,26 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
+        # If next layer's backward computation is done and full params are freed,
+        # no need to prefetch the full params again.
+        # Only prefetch full params if any of the next layer's outputs requires grad
+        if (
```
nit: same comment about the helper function

zhaojuanmao replied:
sounds good

zhaojuanmao commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -595,9 +642,26 @@ def _pre_backward_hook(*unused: Any) -> None:
             if self._is_root:
                 self._queue_wait_for_post_backward()
 
+            if self.backward_prefetch == BackwardPrefetch_.BACKWARD_PRE:
```
we still need to call _rebuild_full_params on L653 even for BackwardPrefetch_.BACKWARD_PRE, which is for fetching full parameters of current layer. e.g., first layer's full parameters may have not been prefetched yet, this call is needed for that case. also, I feel it is more safe to call it in general, if full parameters have been prefetched, _rebuild_full_params will be a no-op

zhaojuanmao commented on torch/distributed/_fsdp/fully_sharded_data_parallel.py:
```diff
@@ -595,9 +642,26 @@ def _pre_backward_hook(*unused: Any) -> None:
+
+            # Prefetch next layer's full params in backward pass,
+            # since it is prefetching, no need to wait for all_gather stream.
+            if (
```
sounds good

zhaojuanmao committed:
enable backward pass computation and communication overlap by prefetching all gather (#70235)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/70235

address comments in https://github.com/pytorch/pytorch/pull/69282:
Have fixed a few corner cases for prefetching full parameters in post backward hook.

After benchmarking, prefetching full parameters in the pre-backward hook has the best performance and stable but at cost of increased memory; prefetching full parameters in the post-backward hook did not see expected performance, also failed in a few corner cases (fixed) although there is no memory increase. The main issue is that post backward hook fire order is not consistent with opposite of forward computation order, so incorrectly prefetched all gather could delay the really needed all gather in the single NCCL stream and cause some layer's computation delay.

So putting  these two algorithms as two configurable experimental algorithms for now

prefetch full parameters at pre-backward hook:

It is observed from past traces that all gather ops are not triggered until current layer's backward pass starts to compute, also for some models previous layers' reduce scatter is scheduled before next layer's all gather ops, since all gather and reduce scatter are in the same nccl stream, this case could result in backward pass has no communication and computation overlap.

To explicitly make next layers' all gather scheduled while previous layers' backward computation is running, we can prefetch next layers' all gather full params. This can help 1) both all gather and reduce scatter are overlapped with computation deterministically 2) only prefetch one layer's all gather full parameters, to avoid increasing too much memories.

The implementation borrowed the idea from facebookresearch/fairscale#865, where forward graph order is recorded in the forward pass.

In the backward pass, this PR prefetches all gather full parameter in current layer's pre-backward hook, instead of prefetching in current layer's post backward hook in facebookresearch/fairscale#865. Also make sure all gather streams are synced properly.

Experiments showed 10% memory increase and 20% latency speed up for 1GB roberta model in a slow network environment.

Test Plan: unit tests

Reviewed By: rohan-varma

Differential Revision: D33252795

fbshipit-source-id: c2683ee7da70ef9b48fd0c03e1d1077a7839f7db
